### PR TITLE
update cccl to v2.3.2

### PR DESCRIPTION
**Pull request by danpetry**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

The versions in cccl-config-version.cmake in the cuda_cccl v12.4.127 tarball for linux shows a version of 2.3.2

### CODE REVIEW

**Comment by danpetry** [comment]
@conda-forge-admin, please rerender

**Comment by jakirkham** [comment]
Thanks Dan! 🙏

Is there a test that we could add to make sure we catch this in the future?

**Comment by adibbley** [comment]
The [release notes](https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html) are still claiming 2.3.1.

**Comment by bdice** [comment]
This shouldn't be changed here, if the release notes are correct. This change has already been made in https://github.com/conda-forge/cccl-feedstock/pull/13, which is in the `cccl` package. The `cuda-cccl` feedstock is meant to track the last CCCL released as a part of the CUDA Toolkit. While 2.3.2 is out, it is not yet shipped in a CTK release.

Please see https://github.com/NVIDIA/cccl?tab=readme-ov-file#conda for a description of what each of the CCCL conda packages is used for.

**Comment by bdice** [comment]
Oops. Just reread the description and saw the comment:

> The versions in cccl-config-version.cmake in the cuda_cccl v12.4.127 tarball for linux shows a version of 2.3.2

If that's correct, then maybe there is work to be done here. I was trusting the release notes, too.

**Comment by jakirkham** [comment]
@jrhemstad do you know what version of CCCL we should expect here?

Is `cccl-config-version.cmake` the best way to validate that?

**Comment by danpetry** [comment]
Something like this?

**Comment by danpetry** [comment]
This seems to be working, and the PR also includes upstreaming of a windows test we put into our recipe

File: recipe/meta.yaml
Comment (bdice) [inline recipe/meta.yaml]: Do you think this would be clearer? (I tested the individual commands but not in a full conda-build)
```suggestion
    - CCCL_VERSION_FILE="$PREFIX/targets/{{ target_name }}/lib/cmake/cccl/cccl-config-version.cmake"  # [linux]
    - CCCL_VERSION_MAJOR=$(sed -n 's/set(CCCL_VERSION_MAJOR \([0-9]*\))/\1/p' ${CCCL_VERSION_FILE})  # [linux]
    - CCCL_VERSION_MINOR=$(sed -n 's/set(CCCL_VERSION_MINOR \([0-9]*\))/\1/p' ${CCCL_VERSION_FILE})  # [linux]
    - CCCL_VERSION_PATCH=$(sed -n 's/set(CCCL_VERSION_PATCH \([0-9]*\))/\1/p' ${CCCL_VERSION_FILE})  # [linux]
    - [[ "${CCCL_VERSION_MAJOR}.${CCCL_VERSION_MINOR}.${CCCL_VERSION_PATCH}" == "{{ cccl_version }}" ]] || exit 1  # [linux]
```

Comment (danpetry) [reply recipe/meta.yaml]: yes thanks, this seems safer

**Review by bdice — APPROVED** [review]
LGTM. @jakirkham @adibbley Any other thoughts before merging?

**Review by adibbley — APPROVED** [review]

File: recipe/meta.yaml
Comment (KyleFromNVIDIA) [inline recipe/meta.yaml]: As we discussed offline with @bdice, I'd suggest writing a `cmake -P` script that `include()`s the `cccl-config-version.cmake` file and prints out a machine-friendly version.

**Review by KyleFromNVIDIA — CHANGES_REQUESTED** [review]

File: recipe/meta.yaml
Comment (bdice) [reply recipe/meta.yaml]: I have this locally and will push in a moment. It is a bit cleaner.

Comment (bdice) [reply recipe/meta.yaml]: Updated in 34c0ddd0d88b1077c3d3da0da0bcc5ad81c21b2c.

Comment (danpetry) [reply recipe/meta.yaml]: Thanks

File: recipe/meta.yaml
Comment (KyleFromNVIDIA) [inline recipe/meta.yaml]: Something to consider: instead of passing `CCCL_VERSION_FILE` as an environment variable, you could pass it as a CMake variable with `-D`:

```suggestion
    - '[[ "$(cmake -DCCCL_VERSION_FILE="$PREFIX/targets/{{ target_name }}/lib/cmake/cccl/cccl-config-version.cmake" -P verify-version.cmake 2>&1)" == "{{ cccl_version }}" ]] || exit 1'                      # [linux]
```

File: recipe/verify-version.cmake
Comment (KyleFromNVIDIA) [inline recipe/verify-version.cmake]: If you pass `-DCCCL_VERSION_FILE=...` in `meta.yaml`, you can remove this.

File: recipe/verify-version.cmake
Comment (KyleFromNVIDIA) [inline recipe/verify-version.cmake]: I did some testing and found that this actually prints to `stderr`, not `stdout` - my bad, I should have realized that sooner (though I don't think I've ever personally used the untyped `message()` signature.) Here's what you want:

```suggestion
        execute_process(COMMAND "${CMAKE_COMMAND}" -E echo "${CCCL_VERSION_MAJOR}.${CCCL_VERSION_MINOR}.${CCCL_VERSION_PATCH}")
```

**Review by KyleFromNVIDIA — CHANGES_REQUESTED** [review]
Overall concept looks good, I left one bug fix and one optional suggestion.

File: recipe/meta.yaml
Comment (bdice) [reply recipe/meta.yaml]: My reason for using an env var was just that it split the test command across two lines more cleanly than one long line with an inline variable. Super minor. I’ll adapt according to your suggestion.

File: recipe/verify-version.cmake
Comment (bdice) [reply recipe/verify-version.cmake]: I worked around this using `2>&1`. I can change this but I kind of preferred piping both together since it’s calling cmake in a subshell.

Comment (KyleFromNVIDIA) [reply recipe/verify-version.cmake]: IMO, `2>&1` is for when you just want to pipe the output of a long command into a human-readable log file. When a program has machine-readable output, `stdout` is meant for said output, and `stderr` is meant for human-readable error messages. Piping them together muddies this distinction and potentially injects error messages into places they shouldn't go. I'd rather spawn `cmake -E echo` in a subshell.

**Review by KyleFromNVIDIA — APPROVED** [review]
Approved once the CI failures are fixed

File: recipe/verify-version.cmake
Comment (danpetry) [reply recipe/verify-version.cmake]: out of interest, why not use `message(STATUS ...)`? it's the only thing that goes to stdout and the FATAL_ERRORs go to stderr. Why's a further subprocess necessary?

Comment (danpetry) [reply recipe/verify-version.cmake]: then remove the `2>&1`

Comment (KyleFromNVIDIA) [reply recipe/verify-version.cmake]: `message(STATUS ...)` prints an extra `-- ` at the beginning, which we're trying to avoid (though I suppose we could just filter it out on the bash side.)

File: recipe/verify-version.cmake
Comment (jakirkham) [inline recipe/verify-version.cmake]:
```suggestion
    message(FATAL_ERROR "CCCL_VERSION_FILE variable is not set.")
```

**Review by jakirkham — COMMENTED** [review]
Thanks all! 🙏

Had one minor suggestion below

**Comment by bdice** [comment]
Great teamwork here -- I appreciate the help. I think we should be good to merge once CI passes.

**Comment by jakirkham** [comment]
Agreed. Thanks all! 🙏

**Review by jakirkham — APPROVED** [review]

**Comment by danpetry** [comment]
Indeed, thanks a lot for the refinement!

**Comment by bdice** [comment]
I opened a similar PR to add a test in `cccl-feedstock` here: https://github.com/conda-forge/cccl-feedstock/pull/16

**Pull request merged**
